protocol/consensus/consensus.go

```diff
@@ -0,0 +1,114 @@
+package consensus
+
+import (
+	"github.com/relab/hotstuff"
+	"github.com/relab/hotstuff/core"
+	"github.com/relab/hotstuff/core/eventloop"
+	"github.com/relab/hotstuff/core/logging"
+	"github.com/relab/hotstuff/modules"
+	"github.com/relab/hotstuff/protocol/proposer"
+	"github.com/relab/hotstuff/protocol/voter"
+	"github.com/relab/hotstuff/protocol/votingmachine"
+	"github.com/relab/hotstuff/service/committer"
+)
+
+// Consensus provides a default implementation of the Consensus interface
+// for implementations of the ConsensusImpl interface.
+type Consensus struct {
+	eventLoop *eventloop.EventLoop
+	logger    logging.Logger
+	config    *core.RuntimeConfig
+
+	committer *committer.Committer
+
+	extHandler modules.ExtProposeHandler
+
+	voter         *voter.Voter
+	votingMachine *votingmachine.VotingMachine
+	proposer      *proposer.Proposer
+}
+
+// New returns a new Consensus instance based on the given Rules implementation.
+func New(
+	// core dependencies
+	eventLoop *eventloop.EventLoop,
+	logger logging.Logger,
+	config *core.RuntimeConfig,
+
+	// protocol dependencies
+	extHandler modules.ExtProposeHandler,
+	proposer *proposer.Proposer,
+	voter *voter.Voter,
+	votingMachine *votingmachine.VotingMachine,
+
+	// service dependencies
+	committer *committer.Committer,
+) *Consensus {
+	cs := &Consensus{
+		eventLoop: eventLoop,
+		logger:    logger,
+		config:    config,
+
+		extHandler:    extHandler,
+		proposer:      proposer,
+		voter:         voter,
+		votingMachine: votingMachine,
+
+		committer: committer,
+	}
+	cs.eventLoop.RegisterHandler(hotstuff.ProposeMsg{}, func(event any) {
+		p := event.(hotstuff.ProposeMsg)
+		cs.OnPropose(&p)
+	})
+	return cs
+}
+
+// OnPropose is called when receiving a proposal from a leader and returns true if the proposal was voted for.
+func (cs *Consensus) OnPropose(proposal *hotstuff.ProposeMsg) {
+	block := proposal.Block
+	// ensure that I can vote in this view based on the protocol's rule.
+	err := cs.voter.Verify(proposal)
+	if err != nil {
+		cs.logger.Errorf("failed to verify incoming vote: %v", err)
+		return
+	}
+	// if we can't commit the block yet, don't vote for it.
+	if !cs.committer.TryCommit(block) {
+		return
+	}
+	// even if voting fails, we should be able to go to the next view.
+	newInfo := hotstuff.NewSyncInfo().WithQC(block.QuorumCert())
+	cs.eventLoop.AddEvent(hotstuff.NewViewMsg{
+		ID:       cs.config.ID(),
+		SyncInfo: newInfo,
+	})
+	// try to vote for the block and retrieve its partial certificate.
+	pc, err := cs.voter.Vote(block)
+	// don't send the vote if it failed
+	if err != nil {
+		cs.logger.Errorf("failed to vote: %v", err)
+		return
+	}
+	cs.extHandler.OnPropose(proposal, pc)
```

Comment:
I do not understand what this is doing. I also looked at the interface and none of the comments there help me understand. Probably this needs different names.